vm/src/builtins/object.rs

```diff
@@ -183,12 +183,24 @@ impl PyBaseObject {
 
     /// Return repr(self).
     #[pymethod(magic)]
-    fn repr(zelf: PyObjectRef) -> String {
-        format!(
-            "<{} object at {:#x}>",
-            zelf.class().tp_name(),
-            zelf.get_id()
-        )
+    fn repr(zelf: PyObjectRef, vm: &VirtualMachine) -> String {
+        let class = zelf.class();
+        match class.module(vm).downcast_ref::<PyStr>().map(|m| m.as_str()) {
+            Some(module) if module != "builtins" => {
+                let class_name = class.name();
+                format!(
+                    "<{}.{} object at {:#x}>",
+                    module,
+                    class
+                        .qualname(vm)
```

Comment:
by the document referenced in the previous PR (https://docs.python.org/3/c-api/typeobj.html?highlight=tp_name#c.PyTypeObject.tp_name)

it is `__module__` + `__name__` or just `tp_name`.  Then isn't this `.name()` not `qualname()`?

Reply:
Yes, you're right. I believed myself at the past but I didn't 😅  I turned the return type into `Option<String>` and made it return `None` if the `qualname` is `None`.

Reply:
Hmm, as I understood, you seem to say `tp_name` is `__module__` + `__name__` so `.name()` should be placed instead of `qualname` with the below section:

> For dynamically allocated type objects, this should just be the type name, and the module name explicitly stored in the type dict as the value for key`'__module__`.
> 
> For statically allocated type objects, the tp_name field should contain a dot. Everything before the last dot is made accessible as the `__module__` attribute, and everything after the last dot is made accessible as the `__name__` attribute.

At first, I implemented by following CPython implementation, the below code:

```c
static PyObject *
object_repr(PyObject *self)
{
    PyTypeObject *type;
    PyObject *mod, *name, *rtn;

    type = Py_TYPE(self);
    mod = type_module(type, NULL);
    if (mod == NULL)
        PyErr_Clear();
    else if (!PyUnicode_Check(mod)) {
        Py_DECREF(mod);
        mod = NULL;
    }
    name = type_qualname(type, NULL);
```

And the expression with `module` + `qualname` and another expression with `tp_name`, expect different result. The one with `tp_name` will be used in only `builtins` module.

The reason that we must use `qualname` and can't replace `qualname` with `name` for compatibility, is that it will make result different. The `module` doesn't treat function as module. Let's see with example. I prepared the example code:

```python
def f():
  class A: ...
  print(A.__module__)
  print(A.__name__)
```

There is a function `f` and the `f` function creates dynamic type `A` for each call. The `A.__module__` will have `__main__` and `A.__name__` will have only `A`. And `A.__qualname__` will be `f.<locals>.A`.

Review with Python documentation:

> For dynamically allocated type objects, this should just be the type name,

The `A.__name__` will be just type name, `A`.

> and the module name explicitly stored in the type dict as the value for key '__module__'.

The `A.__module__` will be just `__main__`.

----

P.S. though I already wrote long lines, if I didn't understand your question well, please let me know 😅

Reply:
I am really sorry to make these confusions. I'd better to see CPython code earlier. Thanks a lot.